keps/sig-etcd/4647-cluster-feature-gate/README.md

```diff
@@ -0,0 +1,597 @@
+How do we handle the data change? 
+1. if the etcd server knows this feature, it should be able to handle the data when the feature is on or off. How it handles the data at WAL entry `i` is determined by the cluster feature flag set from raft before WAL entry `i`. 
+1. if the etcd server does not know this feature, i.e. its version is not compatible with this feature, it means the data is not compatible with the storage version. The storage version migration should handle cleaning up older or newly introduced data.
+
+Overall, when developers introduce new data with a feature, they should be careful to include data cleaning in `schema.UnsafeMigrate`. But we prefer not to have a mechanism to disallow such features to be turned off after they are turned on because new features could be buggy, we need to keep the option to be able to turn it off without bring down the whole cluster.
+
+#### Feature Implementation Change Risks
+
+There are cases when a developer wants to change how an existing feature works. This would make things complicated in a mixed version cluster. One member could have the old implementation while another member uses the latest implementation. How can we make sure the cluster works consistently?
+
+Ideally, the developer should introduced a new feature if they want to do major changes. If that is not desirable, similar to [how Kubernetes handles feature gating changes for compatibility version](https://github.com/kubernetes/enhancements/tree/master/keps/sig-architecture/4330-compatibility-versions#feature-gating-changes), we need to keep the implementation history for at least 2 minor versions (allowed version skew) in the code base if the feature change affects data in anyway, like
+```go
+if s.FeatureEnabled(FeatureA) && s.ClusterVersion() <= "3.6" {implementation 1}
+if s.FeatureEnabled(FeatureA) && s.ClusterVersion() >= "3.7" {implementation 2}
+```
+This way the cluster would work consistently because there is a single ClusterVersion across the whole cluster.
+
+## Design Details
+
+On high level, a cluster feature gate would need:
+1. a way to [register new feature gates](#register-new-feature-gates) when a developer adds new features.
+1. a way to [set the feature gates](#set-the-feature-gates) when each member starts up.
+1. a [consensus algorithm](#consensus-algorithm) to determine if a feature is enabled for the whole cluster, even in mixed version scenarios.
+1. an [interface](#interface-change-for-server-code) to query if a feature is enabled for the whole cluster to be used in the server code.
+1. [client APIs](#client-apis-changes) to query if a feature is enabled for the whole cluster.
+1. a way to [remove a feature gate][#feature-removal] when it is no longer useful or have graduated.
+
+### Register New Feature Gates
+
+A feature can be registered as server level feature or cluster level feature, but not both.
```

Comment:
I would recommend to not think cluster feature and server feature as two different creations. Cluster feature is a   feature that enablement is done at cluster level.